xiaogpt/tts/volc.py

```diff
@@ -0,0 +1,134 @@
+from __future__ import annotations
+
+import logging
+import tempfile
+from pathlib import Path
+from typing import Optional
+import json
+import os
+import time
+import base64
+import threading
+import httpx
+
+from volcengine.ApiInfo import ApiInfo
+from volcengine.Credentials import Credentials
+from volcengine.ServiceInfo import ServiceInfo
+from volcengine.base.Service import Service
+
+
+from xiaogpt.tts.base import AudioFileTTS
+from xiaogpt.utils import calculate_tts_elapse
+
+logger = logging.getLogger(__name__)
+
+
+class VolcTTS(AudioFileTTS):
+    def __init__(self, mina_service, device_id, config):
+        super().__init__(mina_service, device_id, config)
+        self.token = get_token(config)
+        self.client = httpx.Client()
+        logger.info("Initializing VolcTTS {self.token}")
+
+    async def make_audio_file(self, query: str, text: str) -> tuple[Path, float]:
+        tts_payload = json.dumps(
+            {
+                "text": text,
+                "speaker": self.config.volc_tts_speaker,
+                "audio_config": {
+                    "format": "mp3",
+                    "sample_rate": 24000,
+                    "speech_rate": 0,
+                },
+            }
+        )
+
+        req = {
+            "appkey": self.config.volc_tts_app,
+            "token": self.token,
+            "namespace": "TTS",
+            "payload": tts_payload,
+        }
+
+        resp = self.client.post("https://sami.bytedance.com/api/v1/invoke", json=req)
+        try:
+            sami_resp = resp.json()
+            logger.info(f"volc sami_resp {resp.status_code}")
+            if resp.status_code != 200:
+                print(sami_resp)
+        except:
+            logger.error(f"Failed to get tts from volcengine with voice=zh {text}")
+
+        if sami_resp["status_code"] == 20000000 and len(sami_resp["data"]) > 0:
+            audio_data = base64.b64decode(sami_resp["data"])
+            with tempfile.NamedTemporaryFile(
+                suffix=".mp3", mode="wb", delete=False, dir=self.dirname.name
+            ) as f:
+                f.write(audio_data)
+
+        return Path(f.name), calculate_tts_elapse(text)
+
+
+## fetch token and save it to file
+## it's aimed to reduce the request to volcengine
+## it'll throw error if token is requested too frequently (more than 1 times per minute)
+def get_token(config):
+    token_file = (
+        "/tmp/volc_token.json"
+        if Path.exists("/tmp/")
+        else Path.joinpath(os.getenv("TEMP"), "volc_token.json")
+    )
```

Comment:
Instead of using temp, I'd recommend saving it into a file under home directory.

Currently we save mi token at `Path.home() / ".mi.token"`